### PR TITLE
Suggest reviewers action

### PR DESCRIPTION
**Pull request by Nyholm**
This PR brings two features: 
1. Someone can make a comment in a PR to say `@carsonbot find me a reviewer please` and a minute later it will reply with a ping to someone.
2. If a PR is opened an had no comments/reviews in 24(?) hours, then we write a comment with a ping to someone. 

This PR works with a Github action that clones the [git-reviewer script](https://github.com/Nyholm/git-reviewer), checks out the source (symfony/symfony or symfony/symfony-docs) and then runs the `git-reviewer`. That will produce a group of users that also have modified the changed files in the PR. 

I ignore FrameworkBundle, all tests and the Changelogs because they are always changed. I also only looking for contributors for the past 2 years. 

When the `git-reviewer` is done, we run the `bin/console app:review:suggest` command. It looks at the contributor lists, removes people (core team) from a block list and removes all people that already have interacted with the PR. 

Then it just writes a comment.

### CODE REVIEW

File: .github/workflows/find-reviewer.yml
Comment (Nyholm) [inline .github/workflows/find-reviewer.yml:72]: This needs to be added before we merge.

File: composer.json
Comment (Nyholm) [inline composer.json:6]: I hope https://github.com/KnpLabs/php-github-api/pull/931 will be merged and released soon.

File: config/services.yaml
Comment (Nyholm) [inline config/services.yaml:45]: This is debug code. Other PRs have added a proper value here.

File: src/Api/Issue/StdErrIssueApi.php
Comment (Nyholm) [inline src/Api/Issue/StdErrIssueApi.php:12]: This class is never used. But it was really useful when testing/debugging, so I kept it in the PR.

File: composer.json
Comment (acrobat) [reply composer.json:6]: @Nyholm merged and I've release v2.16.0!

Comment (Nyholm) [reply composer.json:6]: Awesome. 
Thank you for the quick merge and release.

**Comment by Nyholm** [comment]
This PR is ready for a review

File: src/Service/ReviewerFilter.php
Comment (Nyholm) [inline src/Service/ReviewerFilter.php:38]: Make sure it has a username

Comment (Nyholm) [reply src/Service/ReviewerFilter.php:38]: They are already filtered out in `filterBlocked`

**Comment by javiereguiluz** [comment]
Tobias, thanks for proposing this feature. I know that some communities have something similar. MY only fear it's whether this will be truly useful to add reviewers ... or it will be just annoying and won't improve the current situation. But, the only way to know it is to try it for real, so I'm 👍  to try this.

About the proposed command -> `@carsonbot find me a reviewer please`  These "textual interfaces" are always complex to use ... so maybe we could design it with very simple commands/orders instead of textually correct phrases. E.g. `@carsonbot help`, `@carsonbot find reviewer`

**Comment by Nyholm** [comment]
> MY only fear it's whether this will be truly useful to add reviewers

Yeah.. I am not sure.. We will see =)

> About the proposed command 

I agree. The only hard requirements at the moment is to write `@carsonbot` and `review` on the same line. So most phrases will be okey. 

- `@carsonbot review`
- `@carsonbot find reviewer please`
- `@carsonbot get a review`
- `review @carsonbot`
- etc

File: src/Subscriber/FindReviewerSubscriber.php
Comment (Nyholm) [inline src/Subscriber/FindReviewerSubscriber.php:76]: FYI @javiereguiluz: Here is the "text command" parsing.

Comment (javiereguiluz) [reply src/Subscriber/FindReviewerSubscriber.php:76]: Very nice! It follows the *"be liberal in what you accept"* principle.

Comment (wouterj) [reply src/Subscriber/FindReviewerSubscriber.php:76]: Should we do `preg_match('/\breview\b/', ...)` to prevent `@carsonbot reviewed` from being matched? (e.g. in case someone doesn't understand how to change the status label)

Comment (Nyholm) [reply src/Subscriber/FindReviewerSubscriber.php]: Done. Im using `preg_match` instead. I added a bunch more tests. The same rules apply. If one is using `@carsonbot` and (`review` or `reviewer`) on the same line, then we will trigger a search for a candidate.

File: .github/workflows/find-reviewer.yml
Comment (stof) [inline .github/workflows/find-reviewer.yml:80]: we might also want to ignore the `UPGRADE` file

File: src/Command/SuggestReviewerCommand.php
Comment (stof) [inline src/Command/SuggestReviewerCommand.php:42]: typos. `Type` vs `Typ` and `ether` vs `either`

File: src/Command/SuggestReviewerCommand.php
Comment (stof) [inline src/Command/SuggestReviewerCommand.php:83]: typo `I'm sorry`

**Comment by Nyholm** [comment]
Thank you @stof for the review. I've updated the PR accordingly.

File: .github/workflows/find-reviewer.yml
Comment (weaverryan) [inline .github/workflows/find-reviewer.yml:5]: Does this mean this workflow is called each time a `find-reviewer` webhooks is received? Can you help me understand how this works? thx :)

File: src/Subscriber/FindReviewerSubscriber.php
Comment (weaverryan) [inline src/Subscriber/FindReviewerSubscriber.php:39]: Ok, so when a new PR is opened, or "ready_for_review", then, 20 hours later, we will run the "suggest reviewer" functionality. 20 hours seems fairly fast to me maybe ?

File: .github/workflows/find-reviewer.yml
Comment (ogizanagi) [reply .github/workflows/find-reviewer.yml:5]: Exact. Here are some resources about it you could find useful:
- https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#repository_dispatch
- https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#create-a-repository-dispatch-event

File: src/Subscriber/FindReviewerSubscriber.php
Comment (ogizanagi) [reply src/Subscriber/FindReviewerSubscriber.php:39]: I also wonder what to do with PRs opened with the "Status: Needs Work" label for whatever reasons (or changed with Carson bot). But that's probably the same as opening a Draft PR, we don't want it?

File: .github/workflows/find-reviewer.yml
Comment (stof) [reply .github/workflows/find-reviewer.yml:5]: see https://github.com/carsonbot/carsonbot/pull/113/files#diff-761989cba708b0fb2f270ad4d799a7a955b1c611d9cefa6fde84611e03e516c0R36

The corresponding code was added in a previous PR, which is indeed weird.

File: src/Subscriber/FindReviewerSubscriber.php
Comment (stof) [reply src/Subscriber/FindReviewerSubscriber.php:39]: the status is set to `needs work` when a reviewer asks for changes. Saying we don't want such PRs would be an issue. It means we only accept PRs when they are perfect at first review or when the contributor can change the PR and re-request a review in less than 20 hours after receiving a review (which is putting a lot of burden on them when they contribute in their free time)

File: .github/workflows/find-reviewer.yml
Comment (Nyholm) [reply .github/workflows/find-reviewer.yml:5]: So I've defined a special event id called `find-reviewer`. In the `GithubPullRequestApi::findReviewer()`  that @stof linked, I am dispatching this special event. 

It is basically me saying: "Hey start the specific CI job called `find-reviewer`".

File: src/Subscriber/FindReviewerSubscriber.php
Comment (Nyholm) [reply src/Subscriber/FindReviewerSubscriber.php:39]: @weaverryan Maybe it is too fast, maybe not. Whenever I make a PR I am secretly and eagerly waiting for some kind of feedback. So 20 hours is suuuper long. 

@ogizanagi, I've not been thinking of that.. I should make sure the PR has the correct status. 

A user may open a PR, then change the status them selves. If so, we should not call for a review.

Comment (wouterj) [reply src/Subscriber/FindReviewerSubscriber.php:39]: I think everyone likes to get a review/comment within a couple hours after opening. However, there is a difference between what we think is reasonable and what we like. 

If I think "when does someone start to add a ping or asks on Slack if someone can please take a look at their PR", I think this margin is closer to 5 days - 1 week than to 20 hours.

Comment (Nyholm) [reply src/Subscriber/FindReviewerSubscriber.php:39]: Should we try with 20 hours for the first 10 pings? Just to see what happens. 

Also note that none of us never will be pinged. We will mostly ping symfony contributors that nobody ever pinged directly on github or slack.

File: src/Service/TaskHandler/SuggestReviewerHandler.php
Comment (Nyholm) [inline src/Service/TaskHandler/SuggestReviewerHandler.php:52]: @ogizanagi I've added this logic. 

If the PR does not have status "needs review" we will try again 20 hours later. Just until somebody else has made a comment.

File: src/Subscriber/FindReviewerSubscriber.php
Comment (Nyholm) [reply src/Subscriber/FindReviewerSubscriber.php:39]: FYI: The rust community have a bot that pings potential reviewers as soon as the PR is open. 

```
https://github.com/rust-lang/rust/pull/78714
```

File: src/Service/TaskHandler/SuggestReviewerHandler.php
Comment (ogizanagi) [reply src/Service/TaskHandler/SuggestReviewerHandler.php:52]: 👍  LGTM

File: src/Service/ReviewerFilter.php
Comment (nicolas-grekas) [inline src/Service/ReviewerFilter.php:23]: does the bot have access to the mergers team?

**Comment by nicolas-grekas** [comment]
Does any project use this logic successfully already?
I'm curious about how relevant this will be.
Also: which problem does it solve? Will this solve something that is actually slowing us down right now? That's a honest question :) I feel like things are pretty smooth, but I'm curious to see who this proposal can improve things.
What's the way to verify this provides value? Enable, then wait and see?

File: src/Service/ReviewerFilter.php
Comment (Nyholm) [reply src/Service/ReviewerFilter.php:23]: Im not sure. 

But that would be a great future update to implement.

**Comment by Nyholm** [comment]
> Does any project use this logic successfully already?

Yes, @javiereguiluz showed some examples from the Ruby community. They have bot that suggest a reviewer when the PR is opened. 

> which problem does it solve?

Just two very tiny problems. 
1. If a PR dont get any attention, it will try to ping someone
2. It will ping a person that probably never have been pinged before

> What's the way to verify this provides value? Enable, then wait and see?

Yes, Wait and see. Maybe this is a good idea, maybe it is something we disable before Christmas. Let's try something new.

**Review by nicolas-grekas — APPROVED** [review]
Let's give it a try :)

**Comment by weaverryan** [comment]
Let's see how it works! Thanks very much Tobias!

**Pull request merged**
